p2p/types/request_map.go

```diff
@@ -0,0 +1,78 @@
+package types
+
+import sync "sync"
+
+type RequestMap struct {
+	sync.Mutex
+	maxcap   uint32
+	elements []*Request
+	nonces   uint32
+}
+
+func (rm *RequestMap) Get() *Request {
+	rm.Lock()
+	defer rm.Unlock()
+
+	rm.nonces++
+	nonce := rm.nonces
+	newreq := &Request{nonce: nonce, ch: make(chan Work)}
+	rm.elements = append(rm.elements, newreq)
+	return newreq
+}
+
+func (rm *RequestMap) Find(nonce uint32) (uint32, chan Work, bool) {
+	rm.Lock()
+	defer rm.Unlock()
+
+	var request *Request
+	var exists bool
+	var index int
+
+	for i := 0; i < len(rm.elements); i++ {
+		if rm.elements[i].nonce == nonce {
+			exists = true
+			index = i
+			request = rm.elements[i]
+		}
+	}
+
+	if exists {
+		rm.elements[index] = nil
+		fhalf := rm.elements[:index]
+		shalf := rm.elements[index+1:]
+		rm.elements = append(fhalf, shalf...)
+		return request.nonce, request.ch, exists
+	}
+
+	return nonce, nil, false
+}
+
+func (rm *RequestMap) Delete(nonce uint32) bool {
+	defer rm.Unlock()
+	rm.Lock()
+
+	var exists bool
+	var index int
+
+	for i := 0; i < len(rm.elements); i++ {
+		if rm.elements[i].nonce == nonce {
+			exists = true
+			index = i
+			break
+		}
+	}
+
+	if exists {
+		close(rm.elements[index].ch)
+		rm.elements[index] = nil
+		fhalf := rm.elements[:index]
+		shalf := rm.elements[index+1:]
+		rm.elements = append(fhalf, shalf...)
+	}
+
+	return exists
+}
+
+func NewRequestMap(cap uint) *RequestMap {
+	return &RequestMap{maxcap: uint32(cap), elements: make([]*Request, 0), nonces: uint32(0)}
```

Comment:
Personal NIT (feel free to ignore): I think multi-lining this will make it more readable but its very much just a preference.